src/gui_common/AddWindowReorderingSupportToSiblings.cs

```diff
@@ -0,0 +1,328 @@
+    ///   <para>
+    ///     Ignored when window reordering paths are not empty.
+    ///   </para>
+    /// </remarks>
+    [Export]
+    public int AutomaticWindowReorderingDepth = 5;
+
+#pragma warning disable CA2213
+
+    /// <summary>
+    ///   Pairs containing window reordering nodes and what sibling of theirs is an ancestor of this class.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     Used to pass the connections from windows.
+    ///   </para>
+    /// </remarks>
+    private List<(AddWindowReorderingSupportToSiblings Node, Node Sibling)> windowReorderingAncestors = new();
+
+    /// <summary>
+    ///   A sibling that is an ancestor of a window that is currently on top of others.
+    /// </summary>
+    private Node? topSibling;
+
+    private Node parent = null!;
+
+#pragma warning restore CA2213
+
+    private bool isReady;
+
+    /// <summary>
+    ///   Used to save what node to reorder when a window requires to be reordered.
+    /// </summary>
+    private System.Collections.Generic.Dictionary<CustomWindow, Node> connectedWindows = new();
+
+    /// <summary>
+    ///   Used to save what siblings are part of the reordering system to prevent reordering more than is necessary.
+    /// </summary>
+    private HashSet<Node> connectedSiblings = new();
+
+    /// <summary>
+    ///   Finds window reordering nodes in ancestors, looks for manual paths if set, otherwise uses automatic search to
+    ///   find first one.
+    /// </summary>
+    /// <returns>
+    ///   Pairs containing window reordering nodes and what sibling of theirs is an ancestor of this class.
+    /// </returns>
+    public static IEnumerable<(AddWindowReorderingSupportToSiblings Node, Node Sibling)> GetWindowReorderingAncestors(
+        Node startingNode, int maxSearchDepth, Array<NodePath>? paths)
+    {
+        if (paths is null || paths.Count == 0)
+        {
+            foreach (var windowReorderingSupport in GetWindowReorderingAncestors(startingNode, maxSearchDepth))
+            {
+                yield return windowReorderingSupport;
+            }
+        }
+        else
+        {
+            foreach (var windowReorderingSupport in GetWindowReorderingAncestors(startingNode, paths))
+                yield return windowReorderingSupport;
+        }
+    }
+
+    public override void _ExitTree()
+    {
+        windowReorderingAncestors.Clear();
+        connectedWindows.Clear();
+        connectedSiblings.Clear();
+        isReady = false;
+    }
+
+    public void ConnectWindow(CustomWindow window, Node topNode, bool recursive)
+    {
+        // Make sure to be connected to other window reordering nodes before trying to connect the window
+        Setup();
+
+        if (recursive)
+        {
+            // Connection is recursive, connect to other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.ConnectWindow(
```

Comment:
Reading this bit of code, I now understand more of what the design you are going for. The higher level reordering level is handled by the window registering itself there as well. Just an observation you don't need to do anything about this.